### PR TITLE
Document GADT constructor arguments

### PR DESCRIPTION
**Pull request by newhoggy**
# Changelog

```yaml
- description: |
    Document GADT constructor arguments
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - refactoring    # QoL changes
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Additional context for the PR goes here. If the PR fixes a particular issue please provide a [link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=) to the issue.

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-api/internal/Cardano/Api/Certificate.hs [hunk 1]
```diff
@@ -114,25 +114,29 @@ import           Network.Socket (PortNumber)
...
-       -> Ledger.ShelleyTxCert (ShelleyLedgerEra era)
-       -> Certificate era
-
-     -- Conway onwards
-     -- TODO: Add comments about the new types of certificates
-     ConwayCertificate
-       :: ConwayEraOnwards era
-       -> Ledger.ConwayTxCert (ShelleyLedgerEra era)
-       -> Certificate era
+  -- Pre-Conway
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Certificate.hs:117]: Should we turn it into haddock?

File: cardano-api/internal/Cardano/Api/Certificate.hs [hunk 2]
```diff
@@ -114,25 +114,29 @@ import           Network.Socket (PortNumber)
...
+  --   6. Genesis delegation
+  --   7. MIR certificates
+  ShelleyRelatedCertificate
+    :: ShelleyToBabbageEra era
+    -- ^ Shelley to babbage era witness
+    -> Ledger.ShelleyTxCert (ShelleyLedgerEra era)
+    -- ^ Shelley ledger transaction certificate
+    -> Certificate era
+
+  -- Conway onwards
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Certificate.hs:132]: Same as for the other constructor? I guess it wouldn't be a very informative hadock.

File: cardano-api/internal/Cardano/Api/Certificate.hs (hunk 1 above)
Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/Certificate.hs:117]: I don't understand what this list means.

File: cardano-api/internal/Cardano/Api/Tx/Body.hs [hunk 3]
```diff
@@ -798,12 +801,15 @@ deriving instance Eq   (TxInsCollateral era)
 deriving instance Show (TxInsCollateral era)
 
 data TxInsReference build era where
+  -- ^ No transaction inputs
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Tx/Body.hs:804]:
```suggestion
  -- | No transaction inputs
```

File: cardano-api/internal/Cardano/Api/Tx/Body.hs [hunk 4]
```diff
@@ -914,26 +923,30 @@ prettyRenderTxOut (TxOutInAnyEra _ (TxOut (AddressInEra _ addr) txOutVal _ _)) =
   <> renderValue (txOutValueToValue txOutVal)
 
 data TxReturnCollateral ctx era where
-
+  -- ^ No return collateral
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Tx/Body.hs:926]:
```suggestion
  -- | No return collateral
```

File: cardano-api/internal/Cardano/Api/Tx/Body.hs [hunk 5]
```diff
@@ -914,26 +923,30 @@ prettyRenderTxOut (TxOutInAnyEra _ (TxOut (AddressInEra _ addr) txOutVal _ _)) =
...
     -> TxOut ctx era
+    -- ^ The return collateral
     -> TxReturnCollateral ctx era
 
 deriving instance Eq   (TxReturnCollateral ctx era)
 deriving instance Show (TxReturnCollateral ctx era)
 
 data TxTotalCollateral era where
-
+  -- ^ No total collateral
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Tx/Body.hs:941]:
```suggestion
  -- | No total collateral
```

File: cardano-api/internal/Cardano/Api/Tx/Body.hs [hunk 6]
```diff
@@ -1140,13 +1192,18 @@ deriving instance Show (TxUpdateProposal era)
 
 data TxMintValue build era where
 
-     TxMintNone  :: TxMintValue build era
+  -- ^ No minting
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Tx/Body.hs:1195]:
```suggestion
  -- | No minting
```

File: cardano-api/internal/Cardano/Api/Tx/Sign.hs [hunk 7]
```diff
@@ -288,36 +290,38 @@ instance IsShelleyBasedEra era => HasTextEnvelope (Tx era) where
...
+    -> TxBodyScriptData era
+    -- ^ The info for each use of each script: the script input data, both
+    -- the UTxO input data (called the "datum") and the supplied input
+    -- data (called the "redeemer") and the execution units.
+
+    -> Maybe (L.TxAuxData (ShelleyLedgerEra era))
+    -- ^ The 'L.TxAuxData' consists of one or several things,
+    -- depending on era:
+    -- + transaction metadata  (in Shelley and later)
+    -- + auxiliary scripts     (in Allegra and later)
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Tx/Sign.hs:315]:
```suggestion
    -- - transaction metadata additionally  (in Shelley and later)
    -- - auxiliary scripts additionally     (in Allegra and later)
```
I think the lists can only be with dashes and stars https://hackage.haskell.org/package/haddock-cheatsheet-0.1.0.1/docs/Doc-Haddock.html#g:9
... or you can add plus signs after stars or something.

File: cardano-api/internal/Cardano/Api/Tx/Sign.hs [hunk 8]
```diff
@@ -467,13 +471,18 @@ instance IsShelleyBasedEra era => HasTextEnvelope (TxBody era) where
...
 
 data TxBodyScriptData era where
-     TxBodyNoScriptData :: TxBodyScriptData era
-     TxBodyScriptData
-       :: AlonzoEraOnwardsConstraints era
-       => AlonzoEraOnwards era
-       -> Alonzo.TxDats (ShelleyLedgerEra era)
-       -> Alonzo.Redeemers (ShelleyLedgerEra era)
-       -> TxBodyScriptData era
+  -- ^ No script data
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Tx/Sign.hs:474]:
```suggestion
  -- | No script data
```

**Review by carbolymer — APPROVED** [review]
Nice! Have you tested it with haddock command? For example: https://github.com/IntersectMBO/cardano-api/blob/main/.github/workflows/github-page.yml#L33
(this workflow takes 40mins so it's not triggered on PRs)

File: cardano-api/internal/Cardano/Api/Certificate.hs (hunk 1 above)
Comment (carbolymer) [reply cardano-api/internal/Cardano/Api/Certificate.hs:117]: It's not starting with `-- |` so this comment it's not a haddock comment.

**Comment by palas** [comment]
FYI: Because we now enforce the use of `cabal-gild` to format cabal files and, in order to reduce the amount of hassle, I have cherry-picked the appropriate commit into your PR. This will make the `cabal-gild` CI check pass for the PR. The commit should be discarded automatically as soon as you either rebase or merge the PR (since it is already in the `main` branch). Feel free to discard it if you want, but that will make the new required action to fail until you rebase.

**Comment by palas** [comment]
FYI: I have rebased your branch because we have done changes to the formatting. I have made a copy of the unrebased branch here: [backup/newhoggy/document-gadt-constructor-arguments](https://github.com/IntersectMBO/cardano-api/tree/backup/newhoggy/document-gadt-constructor-arguments)

**Pull request closed without merging**